### DISCUSSION TITLE
Can't there be multiple ConstraintProvider classes?

### DISCUSSION THREAD

**Discussion by kentbill** [TimefoldAI/timefold-solver#1292] [General]
Hi Teams,
I have two ConstraintProvider classes: A_ContraintProvider implements ConstraintProvider and B_ContraintProvider implements ConstraintProvider， I plan to use one of them (through the SolverConfig.winthConstraintExtendeClass() method) to build Solver in different scenarios, but an exception occurs when starting Springboot: **Multiple score calculator classes A-ConstraintProvider, B_ConstraintProvider) that implements ConstraintProvider were found in the classpath.**
In previous versions, this exception did not appear, and I am not sure which version started it. That is to say, there cannot be more than one ConstraintProvider class in the same project?

**Comment by triceo** [comment]
The solver tries to auto-detect a constraint provider from your codebase, so that you don't need to explicitly configure one. If you have multiple of them on the classpath, how is the solver supposed to pick? It doesn't know, and therefore it fails.

If you want to have both on the classpath, you'll have to provide a solver config to specifically select the constraint provider you want the solver to use.

**Comment by TomCools** [comment]
To give some more detail on @triceo's comment: the way we have currently documented this is as follows: https://docs.timefold.ai/timefold-solver/latest/integration/integration#integrationWithSpringBootProperties

## Solver Configuration
Create separate solverConfig.xml files. Load them via configuration properties: 

```
Application config file
timefold.solver.solver1.solver-config-xml=solverConfig1.xml
timefold.solver.solver2.solver-config-xml=solverConfig2.xml
```

```
solverConfig1.xml
<?xml version="1.0" encoding="UTF-8"?>
<solver>
  <scoreDirectorFactory>
    <constraintProviderClass>org.acme.schooltimetabling.solver.TimetableConstraintProvider</constraintProviderClass>
  </scoreDirectorFactory>
</solver>
```

```
solverConfig2.xml
<?xml version="1.0" encoding="UTF-8"?>
<solver>
  <scoreDirectorFactory>
    <constraintProviderClass>org.acme.schooltimetabling.solver.TimetableConstraintProvider2</constraintProviderClass>
  </scoreDirectorFactory>
</solver>
```

## Injecting the solver
When using multiple SolverConfigurations, you need to also add a qualifier to your injection point with the name of the solver. 

```
Name in application properties file
timefold.solver.**solver1**.solver-config-xml=solverConfig1.xml

@Autowired
@Qualifier("**solver1**") 
SolverManager<Timetable, String> solverManager
```

> This currently doesn't work for a **SolutionManager** instance.

**Comment by kentbill** [comment]
Thank you! @triceo and @TomCools .
I think I can also use **ConstraintWeightOverrides** to disable constraints that I don't want to run, right?

[https://github.com/TimefoldAI/timefold-solver/blob/6e0918e097f8be8ca0463f345b78d6aabcffe8bf/docs/src/modules/ROOT/pages/constraints-and-score/constraint-configuration.adoc?plain=1#L202](url)

**Reply by triceo** [reply]
Overrides are a feature of the planning solution, not of Spring integration - if your solution includes the overrides, they will be used.
